### PR TITLE
exanded case search results

### PR DESCRIPTION
**Pull request by snopoke**
## Product Description
Backend portion of https://dimagi-dev.atlassian.net/browse/USH-1445

Allow users to expand case search results by referencing custom case properties e.g. "potential_duplicate_id" points to a case ID.

## Technical Summary
This allows passing a special case search key into case search which will be used to load additional cases into the results. For example:

* cases have a property called `potential_duplicate_id` which holds the value of another case ID
* the `x_commcare_expand_id_property` search key is set to `potential_duplicate_id`
* the search results will include all cases referenced by the `potential_duplicate_id` on cases from the main search results
  * this will also include related & child cases of these cases

Best review by commit.

The last 2 commits are optional and change the name of the 'special' keys.

## Feature Flag
??

### Safety story
Potential to impact case search requests.

### Automated test coverage
This code has reasonable test coverage and I've added some new tests for the changes

### QA Plan
None

### Migrations
NA

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

File: corehq/apps/case_search/utils.py [hunk 1]
```diff
@@ -283,12 +283,17 @@ def get_related_cases(helper, app_id, case_types, cases):
         for _type in types
     ]
 
-    results = []
+    expanded_case_results = []
+    if expand_id_property:
+        expanded_case_results.extend(get_expanded_case_results(helper, expand_id_property, cases))
+
+    results = expanded_case_results
```
Comment (stephherbers) [inline corehq/apps/case_search/utils.py:290]: looks like if the `results` declaration was kept above there wouldn't be a need for `expanded_case_results` and this line-- was that just for clarity?

**Review by esoergel — APPROVED** [review]

File: corehq/apps/case_search/utils.py [hunk 2]
```diff
@@ -283,12 +283,17 @@ def get_related_cases(helper, app_id, case_types, cases):
...
+    top_level_cases = cases + expanded_case_results
     if paths:
-        results.extend(get_related_case_results(helper, cases, paths))
+        results.extend(get_related_case_results(helper, top_level_cases, paths))
```
[the remark this answers is not in the document]
Comment (snopoke) [reply corehq/apps/case_search/utils.py:293]: I mean to comment on that - I'm not sure if this was needed or not.

File: corehq/apps/case_search/models.py [hunk 3]
```diff
@@ -6,20 +6,55 @@
...
 UNSEARCHABLE_KEYS = (
     CASE_SEARCH_BLACKLISTED_OWNER_ID_KEY,
     'owner_id',
     'include_closed',   # backwards compatibility for deprecated functionality to include closed cases
-)
+) + CONFIG_KEYS + tuple(LEGACY_CONFIG_KEYS.values())
+
+
+@attr.s(frozen=True)
+class CaseSearchRequestConfig:
```
[the remark this answers is not in the document]
Comment (snopoke) [reply corehq/apps/case_search/models.py:37]: I've done a refactor here which I think is the best of both worlds. https://github.com/dimagi/commcare-hq/pull/30845

File: corehq/apps/case_search/models.py [hunk 4]
```diff
@@ -6,21 +6,55 @@
...
+from corehq.apps.case_search.exceptions import CaseSearchUserError
 from corehq.util.quickcache import quickcache
 
 CLAIM_CASE_TYPE = 'commcare-case-claim'
 FUZZY_PROPERTIES = "fuzzy_properties"
 CASE_SEARCH_BLACKLISTED_OWNER_ID_KEY = 'commcare_blacklisted_owner_ids'
 CASE_SEARCH_XPATH_QUERY_KEY = '_xpath_query'
-CASE_SEARCH_REGISTRY_ID_KEY = 'commcare_registry'
-CASE_SEARCH_EXPAND_ID_PROPERTY_KEY = 'commcare_expand_id_property'
+CONFIG_KEY_PREFIX = "x_commcare_"
```
Comment (orangejenny) [inline corehq/apps/case_search/models.py:19]: Late thoughts on naming:
* Do the config keys really need a different prefix? We use just "commcare" for `commcare_blacklisted_owner_ids`, `commcare_search_score`, and in custom fields: https://github.com/dimagi/commcare-hq/blob/179a5c166f76f9f8830c8a5072d4a025aa0a0a49/corehq/apps/custom_data_fields/models.py#L24, and also `commcare_search_score` 
* Do you think `_xpath_query` should also use this prefix?

Comment (snopoke) [reply corehq/apps/case_search/models.py:19]: Quick thoughts:
- `commcare_blacklisted_owner_ids` and `_xpath_query` are special search terms
- `commcare_search_score` is something we return in the results set like `commcare_project` so they are quite different

We certainly don't need the prefix, I just thought it made it clearer that these are not search terms.

**Pull request merged**